dot/core/consensus.go

```diff
@@ -0,0 +1,213 @@
+	"github.com/ChainSafe/gossamer/lib/crypto/sr25519"
+
+	log "github.com/ChainSafe/log15"
+)
+
+// newBabeSession creates a new BABE session
+func (s *Service) newBabeSession() (*babe.Session, error) {
+	log.Debug(
+		"[core] initializing BABE session...",
+		"epoch", s.epochNumber,
+	)
+
+	// AuthorityData comes from NextEpochDescriptor within the ConsensusDigest
+	// of the block Digest, which is included in the first block of each epoch
+	authData := s.bs.AuthorityData()
+	if len(authData) == 0 {
+		return nil, fmt.Errorf("authority data not set")
+	}
+
+	newBlocks := make(chan types.Block)
+	s.blkRec = newBlocks
+
+	epochDone := make(chan struct{})
+	s.epochDone = epochDone
+
+	babeKill := make(chan struct{})
+	s.babeKill = babeKill
+
+	keys := s.keys.Sr25519Keypairs()
+
+	// get best slot to determine next start slot
+	bestHash := s.blockState.BestBlockHash()
+	bestSlot, err := s.blockState.GetSlotForBlock(bestHash)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get slot for latest block: %s", err)
+	}
+
+	// BABE session configuration
+	bsConfig := &babe.SessionConfig{
+		Keypair:          keys[0].(*sr25519.Keypair),
+		Runtime:          s.rt,
+		NewBlocks:        newBlocks, // becomes block send channel in BABE session
+		BlockState:       s.blockState,
+		StorageState:     s.storageState,
+		TransactionQueue: s.transactionQueue,
+		AuthData:         authData,
+		Done:             epochDone,
+		Kill:             babeKill,
+		StartSlot:        bestSlot + 1,
+		SyncLock:         s.syncLock,
+	}
+
+	// create new BABE session
+	bs, err := babe.NewSession(bsConfig)
+	if err != nil {
+		log.Error("[core] failed to initialize BABE session", "error", err)
+		return nil, err
+	}
+
+	log.Debug(
+		"[core] BABE session initialized",
+		"epoch", s.epochNumber,
+	)
+
+	return bs, nil
+}
+
+// handleConsensusDigest checks if the provided header is the block header for
+// the first block of the current epoch, finds and decodes the consensus digest
+// item from the block digest, and then sets the epoch data for the next epoch
+func (s *Service) handleConsensusDigest(header *types.Header) (err error) {
+
+	// check if block is from current epoch
+	err = s.verifyCurrentEpoch(header)
+	if err != nil {
+		return fmt.Errorf("block is not from current epoch")
+	}
+
+	// check if block header digest items exist
+	if header.Digest == nil || len(header.Digest) == 0 {
+		return fmt.Errorf("header digest is not set")
+	}
+
+	// declare digest item
+	var item types.DigestItem
+
+	// decode each digest item and check its type
+	for _, digest := range header.Digest {
+		item, err = types.DecodeDigestItem(digest)
+		if err != nil {
+			return err
+		}
+
+		// check if digest item is consensus digest type
+		if item.Type() == types.ConsensusDigestType {
+
+			// check if first block already set (consensus digest already set)
+			err = s.verifyFirstBlock()
+			if err == nil {
+				return errors.New("failed to set first block: first block already set")
+			}
```

Comment:
this is confusing, verifyFirstBlock doesn't error if the first block is already set, thenn you check for the absense of an error? I think it should be the other way around, just check if firstBlock != nil (and set it to nil when the epoch changes)

Reply:
this also doesn't check for the case where first block is already set, and we receive another block with a lower number that is also in the current epoch

Reply:
fixed